pyface/data_view/data_models/array_data_model.py

```diff
@@ -0,0 +1,298 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+""" Provides an N-dimensional array data model implementation.
+
+This module provides a concrete implementation of a data model for an
+n-dim numpy array.
+"""
+from collections.abc import Sequence
+
+from traits.api import Array, HasRequiredTraits, Instance, observe
+
+from pyface.data_view.abstract_data_model import AbstractDataModel
+from pyface.data_view.abstract_value_type import AbstractValueType
+from pyface.data_view.value_types.api import (
+    ConstantValue, FloatValue, IntValue, TextValue, no_value
+)
+from pyface.data_view.index_manager import TupleIndexManager
+
+
+class _AtLeastTwoDArray(Array):
+    """ Trait type that holds an array that at least two dimensional.
+
+    This calls numpy.atleast_2d during validation to ensure that the value is
+    good.
+    """
+
+    def validate(self, object, name, value):
+        from numpy import atleast_2d
+        value = super().validate(object, name, value)
+        return atleast_2d(value)
+
+    def _default_for_dtype_and_shape(self, dtype, shape):
+        """ Invent a suitable default value for a given dtype and shape. """
+        from numpy import zeros
+
+        if shape is None:
+            value = zeros((0, 0), dtype)
+        else:
+            size = []
+            for item in shape:
+                if item is None:
+                    item = 0
+                elif isinstance(item, Sequence):
+                    # Given a (minimum-allowed-length, maximum-allowed_length)
+                    # pair for a particular axis, use the minimum.
+                    item = item[0]
+                size.append(item)
+            value = zeros(size, dtype)
+        return value
```

Comment:
If I subclassed from `AbstractArray` I'd end up copying the one method from `Array`.  `Array` is not particularly designed to be anything, if you look at the implementation.

And no, we definitely don't want to be setting the `shape` in this case.  This is not a public class intended to be used by anybody else outside this module, any where we are using it we want to be able to accept _any_ shaped array.  If we did this we would be forcing people to use 2D arrays, and there's better ways of expressing that (eg. just using `Array(shape=(None, None))`).

This is a custom trait type designed to be used for one particular trait on one leaf class to accept the sorts of values that we expect, because there currently isn't a way in the `Array` trait to express "at least dimension N".  Should `Array` grow that in the future I'd happily ditch this trait.

Reply:
Yes, you are right we cannot fix the length of the shape.
Maybe something like this:
```
class _AtLeastTwoDArray(AbstractArray):
    def __init__(self, **kwargs):
        default_value = kwargs.pop("default_value", None)
        default_shape = kwargs.get("shape", (0, 0))
        if default_value is None:
            default_value = np.zeros(default_shape)
        super().__init__(default_value=default_value, **kwargs)
    
    def validate(self, object, name, value):
          # the code you have here...

```


Could we please don't override the private method? It would not have been allowed if this wasn't Python.  I would actually prefer sticking with the original `Array` than overridden a private method of a leaf class, and open an issue on traits to allow a minimum number of dimensions to be set for `Array`.

Reply:
(`AbstractArray` uses a different name for the default value, it is `value`, instead of `default_value`.)